src/main/java/duke/task/Deadline.java

```diff
@@ -0,0 +1,118 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * Represents a task with a deadline.
+ */
+public class Deadline extends Task {
+
+    private String deadlineBy;
+
+    private LocalDate deadlineDate;
+
+    private LocalDateTime deadlineDateTime;
+
+    /**
+     * Constructs a Deadline object with the task description and deadline.
+     *
+     * @param description Description of the task.
+     * @param deadlineBy Deadline of the task.
+     */
+    public Deadline(String description, String deadlineBy) {
+        super(description);
+        this.deadlineDate = parseDate(deadlineBy);
+        this.deadlineDateTime = parseDateTime(deadlineBy);
+        this.deadlineBy = deadlineBy;
+    }
+
+    /**
+     * Formats the string representation of deadline object to write to the file.
+     *
+     * @return String representation of the deadline object to be written to the file.
+     */
+    @Override
+    public String toFileString() {
+        String type = "D";
+        return type + " | " + (isDone() ? "1" : "0") + " | " + this.description + " | " + this.deadlineBy;
+    }
+
+    /**
+     * The string representation of deadline object.
+     *
+     * @return String representation of the deadline object.
+     */
+    @Override
+    public String toString() {
+        if (deadlineDate != null) {
+            String formattedDate = deadlineDate.format(DateTimeFormatter.ofPattern("MMM dd yyyy"));
+            return String.format("[D]%s (by: %s)", super.toString(), formattedDate);
+        } else if (deadlineDateTime != null) {
+            String formattedDate = deadlineDateTime.format(DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm"));
+            return String.format("[D]%s (by: %s)", super.toString(), formattedDate);
+        } else {
+            return String.format("[D]%s (by: %s)", super.toString(), deadlineBy);
+        }
+
+    }
+
+    /**
+     * Parses a String to a LocalDate object
+     *
+     * @param date String representation of potential date
+     * @return A LocalDate object if string can be parsed, null otherwise
+     */
+    private LocalDate parseDate(String date) {
+
+        List<DateTimeFormatter> formatters = new ArrayList<>();
+        //List of accepted data formats
+        formatters.add(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+        formatters.add(DateTimeFormatter.ofPattern("dd/MM/yyyy"));
+        formatters.add(DateTimeFormatter.ofPattern("MM/dd/yyyy"));
+        formatters.add(DateTimeFormatter.ofPattern("d/M/yyyy"));
```

Comment:
Good use of formatters list to specify different formattings of date time.
Perhaps consider making this an additional Date Parser class to be used globally?